Remote mongodb (#32)
ni/systemlink-operations-handbook: handbook/data-stores/data-stores.md

> # MongoDB

Author: I am going to try and rename this via a git command once I'm done with general feedback. I want to try and not lose the history on this file.

> MongoDB is used by the majority of SystemLink services as the primary datastore. MongoDB Community Edition is included in the default installation of SystemLink will launch when the SystemLink server boots This is referred to as a *single node* deployments. MongoDB can also be hosted on a separate stand-alone server, a replica set of 3 or more servers, or accessed through MongoDB's Platform as a service offering: MongoDB Atlas. This is referred to as a *multi node* deployment.

Author: Done. Linking to their website in the first mention.

## Assumptions and Prerequisites

Before you configure a remote MongoDB to use with SystemLink, ensure you have the following

> - A server running SystemLink 2021R1 or greater. Refer to Installing and Configuring SystemLink Server and Clients for the basics of setting up a SystemLink server.

Author: Thanks for the reminder. I meant to include this link.

> - A standalone server running MongoDB, multiple servers hosting a MongoDB replica set, or a MongoDB Atlas Account with a user with the `dbAdmin` or similar role. Refer to [Role-Based Access Control](https://docs.mongodb.com/manual/core/authorization/) for details.

Author: Yes, it looks like `readWriteAnyDatabase` will be enough. I've also broken this out into its own bullet for clarity.

|<img src="../../img/mdb-atlas.png" width="1500px"/>| Multi node with MongoDB Atlas | Here SystemLink is hosted on a dedicated server and connects to a MongoDB Atlas cluster. This is recommended for large scale deployments (>200 managed node, 10,000 test steps and results daily). MongoDB Atlas also simplifies database provisioning, operation, backup and restore. These features may be desireable at any scale of deployment. Refer to [MongoDB Atlas](https://www.mongodb.com/cloud/atlas) for details on this service. |

## Single Node Deployments

> By default SystemLink stores all data on databases and the file system hosted on the SystemLink application server. To improve the reliability of data storage and scale to larger volumes of data, remote data stores are recommended for all production deployments.

Reviewer: Do we need to call out both databases and file systems here? Perhaps:

"By default, SystemLink stores all data on the SystemLink application server."

Or a better version of that sentence that doesn't repeat 'SystemLink'.

> By default SystemLink stores all data on databases and the file system hosted on the SystemLink application server. To improve the reliability of data storage and scale to larger volumes of data, remote data stores are recommended for all production deployments.

Author: After this PR I will adding some short chapters on doing this with files and potentially the data finder index. These we be mostly linking out to docs we already have.

> ### Increasing the index cache size for *single node* deployments

Author: Yes we can. Done.

Increase the index cache size to process more data and use the Test Insights or Asset Management modules in a single-node deployment. This avoids high CPU usage that can occur with the default MongoDB memory constraints in SystemLink. NI recommends increasing the available memory for the index to 8GB. The following steps can be used to increase the available memory for the MongoDB index.
  
1. With a user with administrator privileges, log into the desktop of the SystemLink app server.
1. Open **NI SystemLink Server Configuration**.
1. Navigate to **NoSqlDatabase**.
1. Under **The NoSqlServer is launched and managed by SystemLink**, adjust the value in **The cache size of the database in Gigabytes (GB)**.

> 1. Click **Apply**.

Author: Good suggestions. Done.

!!!note
    Refer to [Sizing a SystemLink server](https://www.ni.com/documentation/en/systemlink/latest/setup/sizing-a-systemlink-server/) for guidance on the CPU, memory, and disk needed for SystemLink single node deployments.

> ## Connecting to MongoDB in a Multi Node Configuration

Author: Agreed. Made the titles consistent.

SystemLink supports three types of multi-node configuration. You can configure SystemLink to connect to a standalone instance, replica sets, or MongoDB Atlas.

Connecting SystemLink to a three or more node MongoDB replica set is the recommended configuration for all production deployments.

!!!note

>     While SystemLink can connect and use a sharded MongoDB cluster (`mongos`) it will not take full advantage of the capabilities enabled by sharded clusters.

Author: I've clarified the utility of sharded clusters.

- NI does not support configurations where the MongoDB instance is used by multiple SystemLink servers.
- NI does not support configurations where the MongoDB instance is also used by another production application.

!!!important

>     Due to a bug in the Elixir MongoDB driver, SystemLink cannot connect to MongoDB instances where there is either the `-` or the `_` characters in the MongoDB username provided in the form input or connection string.

Author: I like this suggestion, so I've changed the wording per you feedback.

> When using a stand-alone MongoDB instance, replica set, or Atlas, you have the option to the data stored within the database. This is built into MongoDB Atlas or requires [MongoDB Enterprise Advanced](https://www.mongodb.com/products/mongodb-enterprise-advanced) for stand-alone or replica sets.

Author: I was missing the word encrypt. Good catch

> When using a stand-alone MongoDB instance, replica set, or Atlas, you have the option to the data stored within the database. This is built into MongoDB Atlas or requires [MongoDB Enterprise Advanced](https://www.mongodb.com/products/mongodb-enterprise-advanced) for stand-alone or replica sets.

Author: Missing encryption. Fixed.

> When using a stand-alone MongoDB instance, replica set, or Atlas, you have the option to the data stored within the database. This is built into MongoDB Atlas or requires [MongoDB Enterprise Advanced](https://www.mongodb.com/products/mongodb-enterprise-advanced) for stand-alone or replica sets.

Author: I missed the word encryption. This has been fixed.

### Connecting to Standalone MongoDB instance

> Using a separate server to host MongoDB provides for greater reliability and lower resource utilization for your SystemLink server.

Reviewer 2: ```suggestion
Using a separate server to host MongoDB increases reliability and lowers resource utilization for your SystemLink server.
```

You may use the supplied form input in **NI SystemLink Sever Configuration** when connecting to a server hosting a single standalone MongoDB instance. Refer to [Connecting to a Remote Mongo Database](https://www.ni.com/documentation/en/systemlink/latest/setup/remote-mongo-database/) for steps to connect with this method.

> 

Reviewer: For SystemLink to make use of a Mongo instance, it is necessary to create an admin user in a an authSource database (usually admin). It seems like we should document that process somewhere.

Author: Good call. Included at the beginning of the section on multi node deployments.

### Connecting to MongoDB with Replica Sets

> Replica sets provide a way to create redundancy in your database to mitigate against potential data loss from the loss of a server hosting MongoDB. Refer to MongoDB for a [tutorial on deploying replica sets](https://docs.mongodb.com/manual/tutorial/deploy-replica-set/).

Reviewer 2: I actually like letting people know they're going to MongoDB docs here.

> Replica sets provide a way to create redundancy in your database to mitigate against potential data loss from the loss of a server hosting MongoDB. Refer to MongoDB for a [tutorial on deploying replica sets](https://docs.mongodb.com/manual/tutorial/deploy-replica-set/).

Author: I'm sticking with the original wording per Jenna's feedback.

> Replica sets provide a way to create redundancy in your database to mitigate against potential data loss from the loss of a server hosting MongoDB. Refer to MongoDB for a [tutorial on deploying replica sets](https://docs.mongodb.com/manual/tutorial/deploy-replica-set/).

Reviewer: Could we say "Refer to the MongoDB tutorial on deploying replica sets"?

> To successfully connect to a MongoDB replica set you must provide SystemLink a connection string. If you use the form in **NI SystemLink Configuration**, SystemLink will not take advantage of the redundancy provided by replica sets.

Reviewer 2: ```suggestion
To successfully connect to a MongoDB replica set and create redundancy, you must specify a connection string in NI SystemLink Server Configuration.
```

1. With a user with administrator privileges, log into the desktop of the SystemLink app server.
1. Open **NI SystemLink Server Configuration**.
1. Navigate to **NoSqlDataBase**.
1. Click the **Connect to an externally managed NoSqlDatabase server** radio button.

    !!!note "Secure remote connections"
        NI recommends securing communications between your SystemLink server and MongoDB instance with TLS. For details, refer to [TLS for Remote MongoDB Instances](/network-security/network-security/#tls-for-remote-mongodb-instances).

1. Select the **Use a custom connection string** checkbox.
1. Enter your connection string. [Refer to Connection String URI Format](https://docs.mongodb.com/manual/reference/connection-string/) for details on structuring your connection string.
1. Click **Test Connection** to ensure SystemLink can connect to the MongoDB instance.
1. If the connection test was successful, click **Apply** to restart SystemLink service manager and connect to the MongoDB instance.

!!!important
    Some connection string formats are unsupported or could cause issues in some environments.

>     - Connection strings that contain the query parameter `tls=` are unsupported. Use the `ssl=` query string to achieve the same capability with the same degree of security.

Author: Yes, let's leave it for now and address in a separate PR and address this as a limitation in R1.

>     - The DNS seed list URI format (`mongodb+srv://`) may fail in some environments due to a bug ([Unable to connect to Atlas due to DNS connectivity issues #358](https://github.com/kobil-systems/mongodb/issues/358)) in the Elixir MongoDB driver used by SystemLink. In this case use the [MongoDB standard connection string format](https://docs.mongodb.com/manual/reference/connection-string/#std-label-connections-standard-connection-string-format) to connect your replica set. This affects both self-hosted replica sets and MongoDB Atlas.

Author: I've made the instance where this issue occurs more explicit.

    - When URL escaping characters in your connection string  you must use uppercase characters, e.g. `%2F` not `%2f`.

    Example connection string (line breaks for readability):
    
    ```url
    mongodb://myusername:<password>:
    ec2-123-123-12.compute-1.amazonaws.com27017,
    ec2-234-234-23.compute-1.amazonaws.com:27017,
    ec2-456-456-45.compute-1.amazonaws.com:27017/
    ?authSource=admin
    &replicaSet=rs0
    &readPreference=primary
    &ssl=true
    ```

### Connecting to MongoDB Atlas

> MongoDB Atlas is a fully managed PaaS from MongoDB. Atlas provides for simpler setup and adminstration of a MongoDB replica set than self hosting.

Reviewer 2: ```suggestion
Connect to MongoDB Atlas for simpler setup and adminstration of a MongoDB replica set compared to self hosting. MongoDB Atlas is a fully managed PaaS from MongoDB. 
```

> If you have not setup an Atlas cluster before, refer to [Getting started with Atlas](https://docs.atlas.mongodb.com/getting-started/). Refer to [Setup Atlas Connectivity](https://docs.mongodb.com/guides/cloud/connectionstring/) for steps to obtain a connection string.

Reviewer: I would recommend just replacing "myFirstDatabase" with admin. The Atlas UI actually tells you to replace that string with a proper database, and myFirstDatabase should never exist so having it in your configuration is just confusing.

> If you have not setup an Atlas cluster before, refer to [Getting started with Atlas](https://docs.atlas.mongodb.com/getting-started/). Refer to [Setup Atlas Connectivity](https://docs.mongodb.com/guides/cloud/connectionstring/) for steps to obtain a connection string.

Author: I've incorporated the suggestion and will reference replacing `myFirstDatabase`.

!!!note "MongoDB Atlas free tier is unsupported by SystemLink"

>     Due to constraints on memory in the Atlas free tier, SystemLink will not successfully connect to a cluster running on the free tier. You must use a paid tier to successfully connect. The lowest performance paid tiers of Atlas are supported by SystemLink. If you need assistance evaluating Atlas with SystemLink please contact NI ([customer.requests@ni.com](mailto:customer.requests@ni.com)) or your local account manager.

Reviewer 2: ```suggestion
    Due to constraints on memory in the Atlas free tier, you must use a paid tier for SystemLink to successfully connect. SystemLink supports the lowest performance paid tiers for Atlas. If you need assistance evaluating Atlas with SystemLink please contact NI ([customer.requests@ni.com](mailto:customer.requests@ni.com)) or your local account manager.
```
Does SystemLink support ONLY the lowest performances ones? Or are we just saying that so people won't think they have to spring for the top-shelf stuff?

Reviewer: Perhaps we should just say that SystemLink supports all paid tiers for Atlas?

Author: Stating we support all paid tiers.

> Because Atlas uses replica sets by default, your SystemLink server could be affected by the bug in the Elixir MongoDB driver used by SystemLink as described in [Connecting to MongoDB with Replica Sets](#connecting-to-mongodb-with-replica-sets). When obtaining a connection string from Atlas use the connection string generated for the Node.js driver version 2.0.14. This provided connection string is fully supported by SystemLink.

Reviewer: The other connection string gotchas also apply to Atlas. Would it be worth pulling them into a separate section that can be linked from both places?

# MongoDB

MongoDB is used by the majority of SystemLink services as the primary datastore. MongoDB Community Edition is included in the default installation of SystemLink will launch when the SystemLink server boots.

!!!note

>     By default SystemLink stores all data on databases and the file system hosted on the SystemLink application server. This is referred to as a *single node* deployment. To improve the reliability of data storage and scale to larger volumes of data, remote data stores are recommended for all production deployments. This is referred to as a *multi node* deployment.

Reviewer 2: > By default SystemLink stores all data on databases and the file system hosted on the SystemLink application server.

By default, both the databases and the file system are hosted on the server, right? Right now it reads like only the file system is. I plan to do a sentence-level review on my next round but wouldn't want to introduce an error here.

Author: Yes, the idea here is to point out that all file storage and all databases are on the same server as SystemLink by default. This isn't our recommended approach for production deployments.  I agree this sentence needs to be improved to clarify that.

>     By default SystemLink stores all data on databases and the file system hosted on the SystemLink application server. This is referred to as a *single node* deployment. To improve the reliability of data storage and scale to larger volumes of data, remote data stores are recommended for all production deployments. This is referred to as a *multi node* deployment.

Reviewer 2: I've been wondering this for a while... does anyone actually use single-node deployment over a long period of time?  Since multi-node is the recommended approach, I wonder how much real estate we should give single-node deployment across all our help content.

Also, we could put this at the start of the "SystemLink with Single Node..." section instead, since there's nothing the user does with this info unless they're reading that part. And that'd still be a pretty prominent location.

I really like this differentiation between single-node and multi-node deployment. I'd like to update R2 docs to match this terminology.

Author: So, up till now I expect the single node deployment to be the most common. Not because its the best approach, but rather that (up until recently) it was all we could do. I honestly thing we should be pushing the single node deployment case as eval only. 

> Also, we could put this at the start of the "SystemLink with Single Node..." section instead, since there's nothing the user does with this info unless they're reading that part. And that'd still be a pretty prominent location.
I agree let's hammer that point home. Its clear with this feedback that the hierarchy of the document needs to be adjusted as well.

Reviewer: This is definitely in the "things doctors tell you to do" space of recommended, but also a lot harder, so not everyone is going to actually follow the advice.

## Assumptions and Prerequisites

> 

Reviewer 2: Can give this a bit more structure and help the flow if we put a short sentence here introducing this list. Something like 'Before you configure a remote MongoDB to use with SystemLink, ensure you have the following:'

Author: Great feedback. I will add this as a lead in

- A server running SystemLink 2021R1 or greater. Refer to Installing and Configuring SystemLink Server and Clients for the basics of setting up a SystemLink server.
- A standalone server running MongoDB, multiple servers hosting a MongoDB replica set, or a MongoDB Atlas Account with a user with the `dbAdmin` or similar role. Refer to [Role-Based Access Control](https://docs.mongodb.com/manual/core/authorization/) for details.

> ### SystemLink with Single Node vs Multi Node MongoDB Deployments

Reviewer 2: I love this table! Anything we can do to ease the constant decisions users have to make will always be helpful.

I can see this part living in SPD in the future, since it's ostensibly helping people choose whether they even need to refer to the rest of this documentation (which covers the multi-node use cases only). We could point people here for the multi-node use cases if we still consider them advanced/niche.

Author: I totally agree. I'm thinking about having versions of this table in the other datastore documents to show the relationship. In an omni-bus SPD version I would encourage showing the single node version and have the multi node versions show all the different data stores on separate servers all at once rather than just showing one data store at a time on a separate server (this would be reflective of our recommended approach). Said another way we probably don't need a table with 15+ rows. I bet we could get the point across in 5 or 6.

The following table summarizes when to consider the various supported configurations of MongoDB with SystemLink.

| Diagram | SystemLink + MongoDB Configuration | When to use |
| | ----------- | ----------- |

> |<img src="../../img/single-box.png" width="1500px"/>| Single node with Defaults     | Here SystemLink and MongoDB are hosted on the same server. Use this configuration when evaluating SystemLink or in small deployments of less than 10 managed nodes without the Asset Manager and Test Insights modules. |

Reviewer 2: For each of these "When to use" descriptions, more user-centered language and bulleted lists could help make this clearer and more scannable.

So here, that could look something like:

- You have one server hosting both MongoDB and SystemLink
- You're evaluating SystemLink
- You are working with less than 10 managed nodes
- You do not need the Asset Manager or Test Insights modules.

Author: I agree and sadly we can't do bulleted lists in tables in markdown 😫. Regardless of this constraint I bet these paragraphs can be cleaned up and made more succinct.

> |<img src="../../img/single-box.png" width="1500px"/>| Single node with increase index cache   | Here SystemLink and MongoDB are hosted on the same server. Use this configuration when evaluating SystemLink or in moderate deployments of less than 25 managed nodes and moderate usage of the Test Insights module producing less than 100 test steps and results each day.     |

Reviewer 2: >Single node with increase index cache |  Here SystemLink and MongoDB are hosted on the same server.

This use case confuses me. Introductory note implies that using MongoDB always means going the multi-node deployment route. Consider either adding some more context to that first note where we define single-node deployment or adding that here.

Author: So I decided to include a discussion of the index cache after seeing a recent post on Teams from Josh where he was recommending that and realizing we didn't already have this workflow in SPD. Its worthwhile to document for the folks who aren't in a position to move to a multi-node deployment.

|<img src="../../img/mdb-atlas.png" width="1500px"/>| Multi node with MongoDB Atlas | Here SystemLink is hosted on a dedicated server and connects to a MongoDB Atlas cluster. This is recommended for large scale deployments (>200 managed node, 10,000 test steps and results daily). MongoDB Atlas also simplifies database provisioning, operation, backup and restore. These features may be desireable at any scale of deployment. Refer to [MongoDB Atlas](https://www.mongodb.com/cloud/atlas) for details on this service. |

### Changing the index cache size for *single node* deployments

> To constrain the amount of memory used by MongoDB, the default installation of SystemLink caps the available memory for the database's index at 1GB. This can lead to high CPU usage if the index grows beyond 1GB and must be paged to disk. If you are using the Test Insights or Asset Management modules, NI recommends increasing the available memory for the index to 8GB. The following steps can be used to increase the available memory for the MongoDB index.

Reviewer 2: > If you are using the Test Insights or Asset Management modules, NI recommends increasing the available memory for the index to 8GB.

This seems like the heart of the user goal here, so we should lead with that and then back it up with context/back-end explanation. People would want to increase the index cache size so they can process more data (due to those 2 modules or something else) while still being single node, right?

So maybe a title/first sentence combo along these lines:
Increasing available memory for single-node deployments

Increase the index cache size to process more data and use the Test Insights or Asset Management modules in a single-node deployment. This avoids high CPU usage that can occur with the default (MongoDB?) memory constraints in SystemLink. NI recommends...

Author: Great feedback. I've made these changes.

> 1. With a user with administrator privileges, log into the desktop of the SystemLink app server.

Reviewer 2: Steps look great except they're all step 1 (in this view, at least).

Author: That's a common practice in Markdown. They will render as an ordered list in the handbook, but it makes the entire list more maintainable while also obeying our linting rules.

1. Open **NI SystemLink Server Configuration**.
1. Navigate to **NoSqlDatabase**.
1. Under **The NoSqlServer is launched and managed by SystemLink**, adjust the value in **The cache size of the database in Gigabytes (GB)**.
1. Click **Apply**.

!!!note
    Refer to [Sizing a SystemLink server](https://www.ni.com/documentation/en/systemlink/latest/setup/sizing-a-systemlink-server/) for guidance on the CPU, memory, and disk needed for SystemLink single node deployments.

### Connecting to MongoDB in a Multi Node Configuration

> SystemLink supports connecting to standalone MongoDB instances and replica sets. While SystemLink can connect and use a sharded MongoDB cluster (`mongos`) it will not take full advantage of the capabilities enabled by sharded clusters.

Reviewer 2: Since this is introducing the next 3 subsections, something more direct/overview-y could help orient people here. 'SystemLink supports three types of multi-node configuration. You can configure SystemLink to connect to a standalone instance, replica sets, or MongoDB Atlas.'

Author: Good call. I've included this overview statement.

- NI does not support configurations where the MongoDB instance is used by multiple SystemLink servers.
- NI does not support configurations where the MongoDB instance is also used by another production application.

> SystemLink makes available form input and connection strings to establish a connection between your SystemLink server and MongoDB. Form input may be used when connecting to a single standalone instance, and a connection string must be provided for replica sets and MongoDB Atlas.

Reviewer 2: I don't think we need this -- the steps here cover it, and for the part that's in SPD, I see we use "custom connection string" right now, but can change that.

!!!note "Secure remote connections"
    NI recommends securing communications between your SystemLink server and MongoDB instance with TLS. For details, refer to [TLS for Remote MongoDB Instances](/network-security/network-security/#tls-for-remote-mongodb-instances).

> !!!important

Reviewer 2: It might be worth moving this from this collection of notes into the actual step where they enter the string. We have a lot of sundry notes in this section and people usually don't read all the 'front matter' to a series of steps.

Author: I've moved the TLS note to be inline with the workflow steps.
